cpp/src/stream_compaction/distinct.cu

```diff
@@ -37,62 +32,182 @@
+                                      cuda::thread_scope_device,
+                                      cudf::detail::hash_table_allocator_type>;
 
-  hash_map_type key_map{compute_hash_table_size(num_rows),
+  auto map = static_map{compute_hash_table_size(input.num_rows()),
                         cuco::sentinel::empty_key{COMPACTION_EMPTY_KEY_SENTINEL},
                         cuco::sentinel::empty_value{COMPACTION_EMPTY_VALUE_SENTINEL},
                         detail::hash_table_allocator_type{default_allocator<char>{}, stream},
                         stream.value()};
 
-  auto row_hash = cudf::experimental::row::hash::row_hasher(preprocessed_keys);
-  experimental::compaction_hash hash_key(row_hash.device_hasher(has_null));
+  auto const preprocessed_input =
+    cudf::experimental::row::hash::preprocessed_table::create(input, stream);
+  auto const has_null = nullate::DYNAMIC{cudf::has_nested_nulls(input)};
+
+  auto const row_hasher = cudf::experimental::row::hash::row_hasher(preprocessed_input);
+  auto const key_hasher = experimental::compaction_hash(row_hasher.device_hasher(has_null));
 
-  cudf::experimental::row::equality::self_comparator row_equal(preprocessed_keys);
-  auto key_equal = row_equal.equal_to(has_null, nulls_equal);
+  auto const row_comp  = cudf::experimental::row::equality::self_comparator(preprocessed_input);
+  auto const key_equal = row_comp.equal_to(has_null, nulls_equal);
 
-  auto iter = cudf::detail::make_counting_transform_iterator(
-    0, [] __device__(size_type i) { return cuco::make_pair(i, i); });
-  // insert distinct indices into the map.
-  key_map.insert(iter, iter + num_rows, hash_key, key_equal, stream.value());
+  auto const pair_iter = cudf::detail::make_counting_transform_iterator(
+    size_type{0}, [] __device__(size_type const i) { return cuco::make_pair(i, i); });
+  map.insert(pair_iter, pair_iter + input.num_rows(), key_hasher, key_equal, stream.value());
 
-  auto const output_size{key_map.get_size()};
-  auto distinct_indices = cudf::make_numeric_column(
-    data_type{type_id::INT32}, output_size, mask_state::UNALLOCATED, stream, mr);
-  // write distinct indices to a numeric column
-  key_map.retrieve_all(distinct_indices->mutable_view().begin<cudf::size_type>(),
-                       thrust::make_discard_iterator(),
-                       stream.value());
+  // The output distinct indices.
+  auto output_indices = rmm::device_uvector<size_type>(map.get_size(), stream, mr);
+
+  // If we don't care about order, just gather indices of distinct keys taken from map.
+  if (keep == duplicate_keep_option::KEEP_ANY) {
+    map.retrieve_all(output_indices.begin(), thrust::make_discard_iterator(), stream.value());
+    return output_indices;
+  }
+
+  // Perform reduction on each group of rows compared equal and the results are store
+  // into this array. The reduction operator is:
+  // - If KEEP_FIRST: min.
+  // - If KEEP_LAST: max.
+  // - If KEEP_NONE: count number of appearances.
```

Comment:
Using separate code paths is not easy if not redundant here because they are essentially called in the same functor. Let me rephrase to avoid the "index" word so there will be less confusion.